src/lib.rs

```diff
@@ -0,0 +1,239 @@
+    type Error = &'static str;
+    fn try_from(frame: *mut PyFrameObject) -> Result<Self, Self::Error> {
+        if frame.is_null() {
+            return Err("frame is null");
+        }
+
+        let dframe = unsafe { *frame };
+
+        if dframe.f_code.is_null() {
+            return Err("f_code is null");
+        }
+
+        let code = unsafe { *dframe.f_code };
+        let yielded_coroutine = {
+            (0 < (code.co_flags & CO_COROUTINE)
+                | (code.co_flags & CO_ITERABLE_COROUTINE)
+                | (code.co_flags & CO_ASYNC_GENERATOR))
+                && (!dframe.f_stacktop.is_null())
+        };
+
+        unsafe {
+            let file_name = match code.co_filename.is_null() {
+                true => "null".to_string(),
+                false => CStr::from_ptr(PyUnicode_AsUTF8(code.co_filename))
+                    .to_string_lossy()
+                    .into_owned(),
+            };
+            let func_name = match code.co_name.is_null() {
+                true => "null".to_string(),
+                false => CStr::from_ptr(PyUnicode_AsUTF8(code.co_name))
+                    .to_string_lossy()
+                    .into_owned(),
+            };
+            Ok(FrameData {
+                func_name,
+                file_name,
+                identifier: frame as usize,
+                yielded_coroutine,
+            })
+        }
+    }
+}
+
+/// end can be None due to lack of Return callback
+#[cfg(Py_3_8)]
+struct ProfilerEntry {
+    file_name: String,
+    func_name: String,
+    start: Instant,
+    end: Option<Instant>,
+}
+
+/// Profiler context to be used as a value for ContextVar.
+#[pyclass]
+struct ProfilerContext {
+    entries: std::cell::Cell<HashMap<usize, ProfilerEntry>>,
+}
+
+/// Format an entry into file_name, func_name and duration.
+fn format_entry(entry: &ProfilerEntry) -> (String, String, Option<u128>) {
+    let duration = match entry.end {
+        Some(v) => Some(v.duration_since(entry.start).as_nanos()),
+        None => None,
+    };
+    (entry.file_name.clone(), entry.func_name.clone(), duration)
+}
+
+#[pymethods]
+impl ProfilerContext {
+    #[new]
+    fn new() -> Self {
+        ProfilerContext {
+            entries: std::cell::Cell::new(HashMap::new()),
+        }
+    }
+    #[getter]
+    fn entries(&mut self) -> PyResult<Vec<(String, String, Option<u128>)>> {
+        let mut result = Vec::new();
+        for (_, entry) in self.entries.get_mut().iter() {
+            result.push(format_entry(entry));
+        }
+        Ok(result)
+    }
+}
+
+fn get_context(py: Python, obj: *mut pyo3::ffi::PyObject) -> Option<Py<ProfilerContext>> {
+    let mut context_value: *mut pyo3::ffi::PyObject = std::ptr::null_mut();
+
+    unsafe {
+        match PyContextVar_Get(
+            obj,
+            std::ptr::null_mut(),
+            &mut context_value as *mut *mut pyo3::ffi::PyObject,
+        ) {
+            0 => (),
+            _ => return None,
```

Comment:
I think you might want to handle errors here, but it'll mean changing the function to return `PyResult<Option<...>>:

```suggestion
            -1 => return Err(PyErr::fetch()),
            _ => return None,
```

Reply:
This function is from a callback, so I can't really raise Python errors I believe